File: BaseLibrary/communication/command/handshake/AckCommand.cs
using System.Collections.Generic;
using BaseLibrary.communication.protocol;

namespace BaseLibrary.communication.command.handshake {

    public sealed class AckCommand : AHandShakeCommand, ACommand.Sendable {

        private static readonly List<ISubCommandFactory> subCommandFactories = new List<ISubCommandFactory>();

        public static int RegisterSubCommandFactory(ISubCommandFactory subCommandFactory) {
            int position = subCommandFactories.Count;
            subCommandFactories.Add(subCommandFactory);
            return position;
        }

        public static readonly IFactory<ACommand.Sendable, ACommand> FACTORY = new CommandFactory();
        private sealed class CommandFactory : ACommandFactory {
            internal CommandFactory() : base() { }

            public override bool IsDeserializeable(string s) {
                s = s.Trim();
                if (s.Equals("ACK")) {
                    cache.Cached(s, new AckCommand());
                    return true;
                }
                return false;
            }

            public override bool IsTransferable(ACommand c) {
                if (c is AckCommand) {
                    cache.Cached(c, new AckCommand());
                    return true;
                }
                return false;
            }
        }

        public AckCommand() : base() { }

        string Sendable.Serialize() {
            return "ACK";
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/command/handshake/AckCommand.java

package BaseLibrary.communication.command.handshake;

import BaseLibrary.communication.command.ACommand;
import BaseLibrary.communication.command.ACommandFactory;
import BaseLibrary.communication.protocol.AProtocol;
import BaseLibrary.communication.protocol.IFactory;

public final class AckCommand extends AHandShakeCommand {

	private static final String NAME = "ACK";

	static {
		AProtocol.registerForDeserialize(NAME, AckCommand.class);
	}

	@Override
	public String getCommandName() {
		return NAME;
	}

	public static final IFactory<ACommand, ACommand> FACTORY = new CommandFactory();

	private static final class CommandFactory extends ACommandFactory {
		private CommandFactory() {
		}

		@Override
		public boolean isDeserializeable(String s, AProtocol protocol) {
			s = s.trim();
			if ("ACK".equals(s)) {
				cache.cached(s, new AckCommand());
				return true;
			}
			return false;
		}

		@Override
		public boolean isTransferable(ACommand c) {
			if (c instanceof AckCommand) {
				cache.cached(c, new AckCommand());
				return true;
			}
			return false;
		}
	}

	public AckCommand() {
	}

	@Override
	public String serialize(AProtocol p) {
		return "ACK";
	}
}
